operator/controllers/execution/scans/hook_reconciler.go

```diff
@@ -62,250 +60,232 @@ func (r *ScanReconciler) setHookStatus(scan *executionv1.Scan) error {
-		if err != nil {
-			return err
-		}
-		findingsUploadURL, err := r.PresignedPutURL(scan.UID, "findings.json", defaultPresignDuration)
-		if err != nil {
-			return err
-		}
+	if sErr := r.Status().Update(ctx, scan); sErr != nil {
+		r.Log.Error(sErr, "Unable to update Scan status")
+		return sErr
+	}
 
-		var hook executionv1.ScanCompletionHook
-		if err := r.Get(ctx, types.NamespacedName{Name: nonCompletedHook.HookName, Namespace: scan.Namespace}, &hook); err != nil {
-			r.Log.Error(err, "Failed to get ReadAndWrite Hook for HookStatus")
-			return err
-		}
+	return err
+}
 
-		jobs, err := r.getJobsForScan(scan, client.MatchingLabels{
-			"securecodebox.io/job-type":  "read-and-write-hook",
-			"securecodebox.io/hook-name": nonCompletedHook.HookName,
-		})
-		if err != nil {
-			return err
-		}
-		if len(jobs.Items) > 0 {
-			// Job already exists
-			return nil
-		}
+func (r *ScanReconciler) executeReadOnlyHooks(scan *executionv1.Scan) error {
+	ctx := context.Background()
 
-		jobName, err := r.createJobForHook(
-			&hook,
-			scan,
-			[]string{
-				rawFileURL,
-				findingsFileURL,
-				rawFileUploadURL,
-				findingsUploadURL,
-			},
-		)
+	nonCompletedHooks := getNonCompletedHookPriorityQueue(&scan.Status.ReadOnlyHookStatus)
 
-		// Update the currently executed hook status to "InProgress"
-		err = r.updateHookStatus(scan, executionv1.HookStatus{
-			HookName: nonCompletedHook.HookName,
-			JobName:  jobName,
-			State:    executionv1.InProgress,
-		})
-		return err
-	case executionv1.InProgress:
-		jobStatus, err := r.checkIfJobIsCompleted(scan, client.MatchingLabels{
-			"securecodebox.io/job-type":  "read-and-write-hook",
-			"securecodebox.io/hook-name": nonCompletedHook.HookName,
-		})
-		if err != nil {
-			r.Log.Error(err, "Failed to check job status for ReadAndWrite Hook")
-			return err
-		}
-		switch jobStatus {
-		case completed:
-			// Job is completed => set current Hook to completed
-			err = r.updateHookStatus(scan, executionv1.HookStatus{
-				HookName: nonCompletedHook.HookName,
-				JobName:  nonCompletedHook.JobName,
-				State:    executionv1.Completed,
-			})
-			return err
-		case incomplete:
-			// Still waiting for job to finish
-			return nil
-		case failed:
-			for i, hookStatus := range scan.Status.ReadAndWriteHookStatus {
-				if hookStatus.HookName == nonCompletedHook.HookName {
-					scan.Status.ReadAndWriteHookStatus[i].State = executionv1.Failed
-				} else if hookStatus.State == executionv1.Pending {
-					scan.Status.ReadAndWriteHookStatus[i].State = executionv1.Cancelled
-				}
+	var err error
+	// If nil then all hooks are done
+	if len(nonCompletedHooks) == 0 {
+		r.Log.Info("Marked scan as done as all hooks have completed", "ScanName", scan.Name)
+		scan.Status.State = "Done"
+		var now = metav1.Now()
+		scan.Status.FinishedAt = &now
+	} else if len(nonCompletedHooks) > 0 {
+		for {
+			priorityQueueItem := heap.Pop(&nonCompletedHooks).(*util.PriorityQueueItem)
```

Comment:
The source data structure isn't relevant. Can you use `topNonCompletedHook` or some other more relevant name?

(+ Too little encapsulation shows here as well.)